src/jsonpatcherproxy.js

```diff
@@ -150,176 +90,214 @@ const JSONPatcherProxy = (function() {
-              });
-            } else {
-              instance.defaultCallback({
-                op: 'remove',
-                path: destinationPropKey
-              });
-            }
-            return ownReflectSet(instance, target, key, newValue);
-          } else if (!Array.isArray(target)) {
-            return ownReflectSet(instance, target, key, newValue);
-          }
+      if (revokableInstance && !instance.isProxifyingTreeNow) {
+        revokableInstance.inherited = true;
+      }
+    }
+    // if the new value is an object, make sure to watch it
+    if (
+      newValue &&
+      typeof newValue == 'object' &&
+      !instance.proxifiedObjectsMap.has(newValue)
+    ) {
+      instance.parenthoodMap.set(newValue, {
+        parent: target,
+        path: key
+      });
+      newValue = instance._proxifyObjectTreeRecursively(target, newValue, key);
+    }
+    if (typeof newValue == 'undefined') {
+      if (target.hasOwnProperty(key)) {
+        // when array element is set to `undefined`, should generate replace to `null`
+        if (Array.isArray(target)) {
+          //undefined array elements are JSON.stringified to `null`
+          instance.defaultCallback({
+            op: 'replace',
+            path: destinationPropKey,
+            value: null
+          });
+        } else {
+          instance.defaultCallback({
+            op: 'remove',
+            path: destinationPropKey
+          });
         }
-        if (Array.isArray(target) && !Number.isInteger(+key.toString())) {
-          return ownReflectSet(instance, target, key, newValue);
+        const oldValue = instance.proxifiedObjectsMap.get(target[key]);
+        // was tje deleted a proxified object?
+        if(oldValue) { 
+          instance.parenthoodMap.delete(target[key]);
+          instance.proxifiedObjectsMap.delete(oldValue);
         }
-        if (target.hasOwnProperty(key)) {
-          if (typeof target[key] === 'undefined') {
-            if (Array.isArray(target)) {
-              instance.defaultCallback({
-                op: 'replace',
-                path: destinationPropKey,
-                value: newValue
-              });
-            } else {
-              instance.defaultCallback({
-                op: 'add',
-                path: destinationPropKey,
-                value: newValue
-              });
-            }
-            return ownReflectSet(instance, target, key, newValue);
-          } else {
-            instance.defaultCallback({
-              op: 'replace',
-              path: destinationPropKey,
-              value: newValue
-            });
-            return ownReflectSet(instance, target, key, newValue);
-          }
+        return Reflect.set(target, key, newValue);
+      } else if (!Array.isArray(target)) {
+        return Reflect.set(target, key, newValue);
+      }
+    }
+    if (Array.isArray(target) && !Number.isInteger(+key.toString())) {
+      return Reflect.set(target, key, newValue);
```

Comment:
That means we disregard setting properties to arrays? (if I'm right please add it as a comment, as I'm not 100% sure)

Reply:
You're right. Otherwise, there will be `length` patches all the time.